### PR TITLE
refactor : Handle multiple arguments for intrinsics mapping

### PR DESCRIPTION
**Pull request by shivimathur**
Fixes : https://github.com/lfortran/lfortran/issues/4190

### CODE REVIEW

**Comment by certik** [comment]
Is there some Fortran code that fails in master but works in this PR? If so, let's add it as a test.

**Review by Pranavchiku — COMMENTED** [review]
Yes, we should add tests, you may look to add tests for intrinsic like `isign`, `dsign`, etc.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -829,85 +829,85 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+        {"dmin1", {"min", {"real"}}},
+        {"max0", {"max", {"int4"}}},
+        {"amax0", {"max", {"int4"}}},
+        {"max1", {"max", {"real"}}},
+        {"amax1", {"max", {"real4"}}},
+        {"dmax1", {"max", {"real"}}},
+        {"dcmplx", {"cmplx", {"any"}}},
+        {"dacos", {"acos", {"real8"}}},
+        {"dacosh", {"acosh", {"real8"}}},
+        {"dint", {"aint", {"real8", "real8"}}},
```
Comment (HarshitaKalani) [inline src/lfortran/semantics/ast_common_visitor.h:892]: Does dint have a second argument? Is it `kind`? If it is so the type of kind will be int, right?

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -829,85 +829,85 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+        {"dnint", {"anint", {"real8", "real8"}}},
```
Comment (HarshitaKalani) [inline src/lfortran/semantics/ast_common_visitor.h:893]: Same for dnint

File: src/lfortran/semantics/ast_common_visitor.h (hunk 1 above)
Comment (shivimathur) [reply src/lfortran/semantics/ast_common_visitor.h:892]: Yes, thanks for pointing this out. I missed it.

**Comment by shivimathur** [comment]
GFortran does not support two arguments for ```dint``` and ```dnint```, but LFortran does. 
```fortran
program main 
    print*, dint(1.0_8, 8)
    if (abs(dint(1.0_8, 8) - 1.0_8) > 10e-5 ) error stop
  
    print*, dnint(1.0_8)
    if (abs(dnint(1.0_8) - 1.0_8) > 10e-5 ) error stop

end program 
```

```console
$ gfortran a.f90 && ./a.out
a.f90:3:11:

    3 |     print*, dint(1.0_8, 8)
      |           1
Error: Too many arguments in call to ‘dint’ at (1)
a.f90:4:12:

    4 |     if (abs(dint(1.0_8, 8) - 1.0_8) > 10e-5 ) error stop
      |            1
Error: Too many arguments in call to ‘dint’ at (1)
```
With master branch : 
```console
$ lfortran a.f90
semantic error: Argument 2 of dint must be of double precision real type
 --> a.f90:3:13
  |
3 |     print*, dint(1.0_8, 8)
  |             ^^^^^^^^^^^^^^
```
With this PR : 
```console
$ lfortran a.f90
1.00000000000000000e+00
1.00000000000000000e+00
```

File: integration_tests/CMakeLists.txt [hunk 3]
```diff
@@ -805,12 +805,13 @@ RUN(NAME intrinsics_241 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc) # logical
 RUN(NAME intrinsics_242 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc) # nint
 RUN(NAME intrinsics_243 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc) # sqrt, abs, log (all kind of real)
 RUN(NAME intrinsics_244 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran) # cshift
-RUN(NAME intrinsics_245 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc) #cmplx
+RUN(NAME intrinsics_245 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc) # cmplx
 RUN(NAME intrinsics_246 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran) # adjustl
 RUN(NAME intrinsics_247 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran) # adjustr
-RUN(NAME intrinsics_248 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc NOFAST fotran) # eoshift
+RUN(NAME intrinsics_248 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc NOFAST fortran) # eoshift
```
Comment (certik) [inline integration_tests/CMakeLists.txt:811]: Great catch!

**Comment by certik** [comment]
Is there code that works with GFortran, but fails in master, but works in this PR?

We need to be careful with creating extensions, especially for these old functions.

**Comment by shivimathur** [comment]
> Is there code that works with GFortran, but fails in master, but works in this PR?
> 
> We need to be careful with creating extensions, especially for these old functions.

Not with the existing functions, but ```dbesjn``` and ```dbesyn``` are yet to be implemented. 
```fortran
program main
    use, intrinsic :: iso_fortran_env, only: dp => real64, sp => real32
    integer :: n
    real(dp) :: x, result

    n = 0
    x = 2.0_dp
    result = dbesjn(n, x)
    print *, result
    if (abs(result - 0.22389077914123567_dp) > 10e-5_dp) error stop

    n = 1
    x = 1.0_dp
    result = dbesjn(n, x)
    print *, result
    if (abs(result - 0.440050585744933_dp) > 10e-5_dp) error stop

end program
```
```console
$ gfortran b.f90 && ./a.out
  0.22389077914123567     
  0.44005058574493350    
```
If implemented with the master branch : 
```console
$ lfortran b.f90
semantic error: Argument 2 of dbesjn must be of integer type
  --> b.f90:41:14
   |
41 |     result = dbesjn(n, x)
   |              ^^^^^^^^^^^^ 

```
If implemented with this PR : 
```console
$ lfortran b.f90
2.23890779141235674e-01
4.40050585744933498e-01
```

However, the tests fail in the fortran backend, I am working on it.

**Comment by certik** [comment]
Excellent! So let's write some tests using `dbesjn` and any other function that failed before.

Regarding `dint` and other such functions, I think we need to insert special handling and restrict it to only accept one argument.

File: integration_tests/intrinsics_251.f90 [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    x8 = 1.0_dp
+    print*, dnint(x8, 4)
+    if (abs(dnint(x8, 4) - 1.0_dp) > 10e-5) error stop
+
+    x8 = 2.0_dp
+    print*, dnint(x8, 8)
+    if (abs(dnint(x8, 8) - 2.0_dp) > 10e-5) error stop
+
+end program
```
Comment (HarshitaKalani) [inline integration_tests/intrinsics_251.f90:33]: Yep, kind arg actually doesn't make sense. `dnint` is for real8 only. We need to handle the number of arguments too in this design.

**Review by HarshitaKalani — COMMENTED** [review]
>the tests fail in the fortran backend

It's okay, you can skip that part. Just add a test without the fortran label.

**Review by certik — APPROVED** [review]

**Pull request merged**